beacon_node/network/src/sync/sampling.rs

```diff
@@ -235,57 +245,103 @@ impl<T: BeaconChainTypes> ActiveSamplingRequest<T> {
+                debug!(self.log, "Sample download success"; "block_root" => %self.block_root, "column_indexes" => ?column_indexes, "count" => resp_data_columns.len());
                 metrics::inc_counter_vec(&metrics::SAMPLE_DOWNLOAD_RESULT, &[metrics::SUCCESS]);
 
-                // No need to check data_columns has len > 1, as the SyncNetworkContext ensure that
-                // only requested is returned (or none);
-                if let Some(data_column) = data_columns.pop() {
+                // Filter the data received in the response using the requested column indexes.
+                let mut data_columns = vec![];
+                for column_index in column_indexes {
+                    let Some(request) = self.column_requests.get_mut(column_index) else {
+                        warn!(
+                            self.log,
+                            "Active column sample request not found"; "block_root" => %self.block_root, "column_index" => column_index
+                        );
+                        continue;
+                    };
+
+                    let Some(data_pos) = resp_data_columns
+                        .iter()
+                        .position(|data| &data.index == column_index)
+                    else {
+                        // Peer does not have the requested data.
+                        // TODO(das) what to do?
+                        debug!(self.log, "Sampling peer claims to not have the data"; "block_root" => %self.block_root, "column_index" => column_index);
+                        request.on_sampling_error()?; // TODO: should return error here?
+                        continue;
+                    };
+
+                    data_columns.push(resp_data_columns.swap_remove(data_pos));
+                }
+
+                if !resp_data_columns.is_empty() {
+                    let resp_column_indexes = resp_data_columns
+                        .iter()
+                        .map(|d| d.index)
+                        .collect::<Vec<_>>();
+                    debug!(
+                        self.log,
+                        "Received data that was not requested"; "block_root" => %self.block_root, "column_indexes" => ?resp_column_indexes
+                    );
+                }
+
+                // Handle the downloaded data columns.
+                if data_columns.is_empty() {
+                    debug!(self.log,"Received empty response"; "block_root" => %self.block_root);
+                    self.column_indexes_by_sampling_request
+                        .remove(&sampling_request_id);
+                } else {
+                    // Overwrite `column_indexes` with the column indexes received in the response.
+                    let column_indexes = data_columns.iter().map(|d| d.index).collect::<Vec<_>>();
+                    self.column_indexes_by_sampling_request
+                        .insert(sampling_request_id, column_indexes.clone());
```

Comment:
What's the reason for this insert? 
In this case, the sampling request id should already be in the mapping right?